src/drivers/magnetometer/qmc5883/qmc5883_i2c.cpp

```diff
@@ -0,0 +1,181 @@
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file QMC5883_I2C.cpp
+ *
+ * I2C interface for QMC5883
+ */
+
+/* XXX trim includes */
+#include <px4_config.h>
+
+#include <sys/types.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+#include <assert.h>
+#include <debug.h>
+#include <errno.h>
+#include <unistd.h>
+
+#include <arch/board/board.h>
+
+#include <drivers/device/i2c.h>
+#include <drivers/drv_mag.h>
+#include <drivers/drv_device.h>
+
+#include "qmc5883.h"
+
+#include "board_config.h"
+
+#define QMC5883L_ADDRESS		0x0D
+
+device::Device *QMC5883_I2C_interface(int bus);
+
+class QMC5883_I2C : public device::I2C
+{
+public:
+	QMC5883_I2C(int bus);
+	virtual ~QMC5883_I2C() = default;
+
+	virtual int	read(unsigned address, void *data, unsigned count);
+	virtual int	write(unsigned address, void *data, unsigned count);
+
+	virtual int	ioctl(unsigned operation, unsigned &arg);
+
+protected:
+	virtual int	probe();
+
+};
+
+device::Device *
+QMC5883_I2C_interface(int bus)
+{
+	return new QMC5883_I2C(bus);
+}
+
+QMC5883_I2C::QMC5883_I2C(int bus) :
+	I2C("QMC5883_I2C", nullptr, bus, QMC5883L_ADDRESS, 400000)
+{
+	_device_id.devid_s.devtype = DRV_MAG_DEVTYPE_QMC5883;
+}
+
+int
+QMC5883_I2C::ioctl(unsigned operation, unsigned &arg)
+{
+	int ret;
+
+	switch (operation) {
+
+	case MAGIOCGEXTERNAL:
+		return external();
+
+	case DEVIOCGDEVICEID:
+		return CDev::ioctl(nullptr, operation, arg);
+
+	default:
+		ret = -EINVAL;
+	}
+
+	return ret;
+}
+
+int
+QMC5883_I2C::probe()
+{
+	uint8_t data[2] = {0, 0};
+
+	// must read registers 0x00 once or reset to read ID registers reliably
+	read(0x00, &data[0], 1);
+	read(0x00, &data[0], 1);
+	read(0x00, &data[0], 1);
+
+	_retries = 10;
+
+	unsigned count = _retries;
+	bool read_valid = false;
+	bool id_valid = false;
+
+	while(count > 0){
```

Comment:
I added a number of retries to probe for in cases of poor wiring or i2c interference.  This has helped with the driver not starting on boot.